evidence/store.go

```diff
@@ -0,0 +1,188 @@
+
+"evidence-lookup"/<evidence-height>/<evidence-hash> -> EvidenceInfo
+"evidence-outqueue"/<priority>/<evidence-height>/<evidence-hash> -> EvidenceInfo
+"evidence-pending"/<evidence-height>/<evidence-hash> -> EvidenceInfo
+*/
+
+type EvidenceInfo struct {
+	Committed bool
+	Priority  int
+	Evidence  types.Evidence
+}
+
+const (
+	baseKeyLookup   = "evidence-lookup"   // all evidence
+	baseKeyOutqueue = "evidence-outqueue" // not-yet broadcast
+	baseKeyPending  = "evidence-pending"  // broadcast but not committed
+)
+
+func keyLookup(evidence types.Evidence) []byte {
+	return keyLookupFromHeightAndHash(evidence.Height(), evidence.Hash())
+}
+
+// big endian padded hex
+func be(h int) string {
+	return fmt.Sprintf("%0.16X", h)
+}
+
+func keyLookupFromHeightAndHash(height int, hash []byte) []byte {
+	return _key("%s/%s/%X", baseKeyLookup, be(height), hash)
+}
+
+func keyOutqueue(evidence types.Evidence, priority int) []byte {
+	return _key("%s/%s/%s/%X", baseKeyOutqueue, be(priority), be(evidence.Height()), evidence.Hash())
+}
+
+func keyPending(evidence types.Evidence) []byte {
+	return _key("%s/%s/%X", baseKeyPending, be(evidence.Height()), evidence.Hash())
+}
+
+func _key(fmt_ string, o ...interface{}) []byte {
+	return []byte(fmt.Sprintf(fmt_, o...))
+}
+
+// EvidenceStore is a store of all the evidence we've seen, including
+// evidence that has been committed, evidence that has been verified but not broadcast,
+// and evidence that has been broadcast but not yet committed.
+type EvidenceStore struct {
+	db dbm.DB
+}
+
+func NewEvidenceStore(db dbm.DB) *EvidenceStore {
+	return &EvidenceStore{
+		db: db,
+	}
+}
+
+// PriorityEvidence returns the evidence from the outqueue, sorted by highest priority.
+func (store *EvidenceStore) PriorityEvidence() (evidence []types.Evidence) {
+	// reverse the order so highest priority is first
+	l := store.ListEvidence(baseKeyOutqueue)
+	l2 := make([]types.Evidence, len(l))
+	for i, _ := range l {
+		l2[i] = l[len(l)-1-i]
+	}
+	return l2
+}
+
+// PendingEvidence returns all known uncommitted evidence.
+func (store *EvidenceStore) PendingEvidence() (evidence []types.Evidence) {
+	return store.ListEvidence(baseKeyPending)
+}
+
+// ListEvidence lists the evidence for the given prefix key.
+// It is wrapped by PriorityEvidence and PendingEvidence for convenience.
+func (store *EvidenceStore) ListEvidence(prefixKey string) (evidence []types.Evidence) {
+	iter := store.db.IteratorPrefix([]byte(prefixKey))
+	for iter.Next() {
+		val := iter.Value()
+
+		var ei EvidenceInfo
+		wire.ReadBinaryBytes(val, &ei)
+		evidence = append(evidence, ei.Evidence)
+	}
+	return evidence
+}
+
+// GetEvidence fetches the evidence with the given height and hash.
+func (store *EvidenceStore) GetEvidence(height int, hash []byte) *EvidenceInfo {
+	key := keyLookupFromHeightAndHash(height, hash)
+	val := store.db.Get(key)
+
+	if len(val) == 0 {
+		return nil
+	}
+	ei := new(EvidenceInfo)
```

Comment:
there are at least two cases of `var ei EvidenceInfo` and `&ei` elsewhere however this is the only case of `ei := new(EvidenceInfo)` in this PR. since the variable is named the same maybe it is easier reading to convert this last one to the `var` form also for parallel consistency.